File: Third Year/Semester 5/Formal Languages and Compiler Techniques/Laboratory 2/Lab2/HashTable.cpp
#include <iostream>
#include <vector>
#include <string>

template <typename T>
class HashTable {
private:
    std::vector<std::vector<T>> hashTable;
    int capacity;

public:
    HashTable(int capacity) : capacity(capacity) {
        hashTable.resize(capacity);
    }

    int getCapacity() {
        return capacity;
    }

    int hash(int key) {
        return key % capacity;
    }

    int hash(const std::string& key) {
        int hash = 5381;
        for (char c : key) {
            hash = ((hash << 5) + hash) + c;
        }
        return std::abs(hash) % capacity;
    }

    bool contains(T key) {
        int hashValue = getHashValue(key);
        for (const T& item : hashTable[hashValue]) {
            if (item == key) {
                return true;
            }
        }
        return false;
    }

    int getHashValue(T key) {
        int hashValue = -1;
        if (std::is_same<T, int>::value) {
            hashValue = hash(key);
        } else if (std::is_same<T, std::string>::value) {
            hashValue = hash(key);
        }
        return hashValue;
    }

    void add(T key) {
        int hashValue = getHashValue(key);
        if (!contains(key)) {
            hashTable[hashValue].push_back(key);
        }
    }

    int getPosition(T key) {
        int hashValue = getHashValue(key);
        if (hashValue != -1) {
            return hashValue;
        }
        return -1; // Key not found or not supported
    }

    friend std::ostream& operator<<(std::ostream& os, const HashTable<T>& table) {
        os << "SymbolTable { items=";
        for (const std::vector<T>& bucket : table.hashTable) {
            os << "[";
            for (const T& item : bucket) {
                os << item << " ";
            }
            os << "]";
        }
        os << " }";
        return os;
    }
};

File: Third Year/Semester 5/Formal Languages and Compiler Techniques/Laboratory 2/HashTable.py
class HashTable:
    # Constructor to initialize the HashTable with a given capacity
    def __init__(self, capacity):
        # Create an empty list to represent the hash table with 'capacity' empty buckets
        self.hashTable = [[] for _ in range(capacity)]
        self.capacity = capacity

    def get_capacity(self):
        return self.capacity

    def hash(self, key):
        if isinstance(key, int):
            return key % self.capacity  # Use modulo to get the hash value
        elif isinstance(key, str):
            hash_val = 5381
            for c in key:
                # Update the hash value using a hashing algorithm for strings
                hash_val = ((hash_val << 5) + hash_val) + ord(c)
            return abs(hash_val) % self.capacity

    def contains(self, key):
        # Get the hash value for the key
        hash_value = self.get_hash_value(key)
        for item in self.hashTable[hash_value]:
            if item == key:
                return True
        return False

    # Get the hash value for a given key (integer or string)
    def get_hash_value(self, key):
        hash_value = -1
        if isinstance(key, int):
            hash_value = self.hash(key)
        elif isinstance(key, str):
            hash_value = self.hash(key)
        return hash_value

    # # Add a key to the hash table
    def add(self, key):
        hash_value = self.get_hash_value(key)
        self.hashTable[hash_value].append(key)

    # Get the position (bucket index) for a key
    def get_position(self, key):
        hash_value = self.get_hash_value(key)
        if hash_value != -1:
            return hash_value  # Return the bucket index where the key would be stored
        return -1

    # Return a string representation of the hash table
    def __str__(self):
        return str(self.hashTable)